rust/src/derivation.rs

```diff
@@ -443,3 +444,96 @@ pub unsafe extern "C" fn zcashlc_get_sapling_receiver_for_unified_address(
     });
     unwrap_exc_or_null(res)
 }
+
+/// Derives and returns a ZIP 32 Arbitrary Key from the given seed at the "wallet level", i.e.
+/// directly from the seed with no ZIP 32 path applied.
+///
+/// The resulting key will be the same across all networks (Zcash mainnet, Zcash testnet, OtherCoin
+/// mainnet, and so on). You can think of it as a context-specific seed fingerprint that can be used
+/// as (static) key material.
+///
+/// `context_string` is a globally-unique non-empty sequence of at most 252 bytes that identifies
+/// the desired context.
+///
+/// # Safety
+///
+/// - `context_string` must be non-null and valid for reads for `context_string_len` bytes, and it
+///   must have an alignment of `1`.
+/// - The memory referenced by `context_string` must not be mutated for the duration of the function
+///   call.
+/// - The total size `context_string_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - `seed` must be non-null and valid for reads for `seed_len` bytes, and it must have an
+///   alignment of `1`.
+/// - The memory referenced by `seed` must not be mutated for the duration of the function call.
+/// - The total size `seed_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of pointer::offset.
+/// - Call `zcashlc_free_boxed_slice` to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_derive_arbitrary_wallet_key(
+    context_string: *const u8,
+    context_string_len: usize,
+    seed: *const u8,
+    seed_len: usize,
+) -> *mut FfiBoxedSlice {
+    let res = catch_panic(|| {
+        let context_string = unsafe { slice::from_raw_parts(context_string, context_string_len) };
+        let seed = unsafe { slice::from_raw_parts(seed, seed_len) };
+
+        let key = arbitrary::SecretKey::from_path(context_string, seed, &[]);
```

Comment:
Given that this will panic if `context_string` is greater than 252 bytes, should that be checked?

Also, now that I'm thinking about this, why was this implemented as a panic in https://github.com/zcash/zip32/pull/19/files?

Reply:
I suppose that this is not intended to be a dynamically generated value, so a panic makes sense, because you should be using a constant for the context string, but generally it seems better to use `Result` for this sort of thing, because the caller can always choose to `unwrap`. A user who simply uses e.g. IDE autocomplete to take advantage of a method like `arbitrary::SecretKey::from_path` will miss that it can panic, and this can therefore be a source of avoidable runtime crashes.

Reply:
If it causes a panic then it must be documented in a Panics section of the doc comment. I think it is preferable to have it be an error, though.

Reply:
This does not cause a panic in the Swift SDK. There's a `catch_panic` a few lines above here, that converts the Rust panic into an error.